### DISCUSSION TITLE
Probe Help Needed...

### DISCUSSION THREAD

**Discussion by plazarch** [nebhead/PiFire#117] [Q&A]
I am getting oh so close.  I have the display working, the hopper sensor working, all the relays working, but the probes are driving me insane.  I have set all the probe types to ThermoWorks Pro and double checked in settings.json.  With no probes connected I see a Grill temp of 393 to 394 degrees F, Probe-1 and Probe-2 are at 0 degrees.  If I plug a probe in to the connectors for Grill Probe I get the same grill temp, but Probe-1 shows 77 degrees which is about correct for the ambient temp.  If I plug a probe in to the connectors for Probe-1 I get the same grill temp, 0 degrees for Probe-1 and 77 degrees for Probe-2.  If I plug a probe into Probe-2 I get the same grill temp and 0 degrees on Probe-1 and Probe-2.  I have double checked solder joints and probed the connections between the screw connectors and the ADS1115.  Everything appears to be correct.  Grill Probe connects to A0, Probe-1 connects to A1, and Probe-2 connects to A2.  It almost seems like something is shifted by one but I haven't been able to figure out where.  I have enabled debug and found nothing of interest in the logs.  No errors other than the distance sensor which currently is connected.  With it connected I get the same behavior on the probes.  I've got v2.00a mainboard, weberbox's button board, a known good Raspberry Pi Zero 2 W.  Please help me nebhead, you're my only hope! :-)

![PXL_20220826_163631776](https://user-images.githubusercontent.com/16919300/186954339-8eba0cc6-44db-404f-a1a2-090e2bfc463a.jpg)
No Probe Connected

![PXL_20220826_163555652](https://user-images.githubusercontent.com/16919300/186954412-20edb90e-26c4-40ef-85bb-a19f6518ff18.jpg)
Probe Connected to Grill_Probe Connector

![PXL_20220826_163613651](https://user-images.githubusercontent.com/16919300/186954512-e5d67ffd-c294-4f07-bd73-c2faa9122f3a.jpg)
Probe Connected to Probe_1 Connector

**Comment by plazarch** [comment]
Well, after walking away from it for a while, eating lunch, and having a litle siesta, I came back to it and changed settings.json to be:
<code>    "probe_sources": [
      "ADC1",
      "ADC2",
      "ADC3"
    ],
</code> and all is working normally.  Not sure why I need to do this, but it is working for now.

**Reply by nebhead** [reply]
Very interesting.  Are you using the configuration with the MAX31865?  That is the only thing that I can think would adjust the probe_sources any other way.

**Comment by plazarch** [comment]
No, I am/was not using the MAX31865.  I also had another ADS1115 from another project, but it was a hat style so I worked that in and wired up probes and had the exact same issue, so it was something with the software.  I'm just glad I found a workaround as my wife was starting to question if it would've been cheaper to by the wifi Traeger model. :-(  Now I can tell her that definitely wouldn't have been cheaper and I have more functionality. :-)

**Reply by nebhead** [reply]
> No, I am/was not using the MAX31865. I also had another ADS1115 from another project, but it was a hat style so I worked that in and wired up probes and had the exact same issue, so it was something with the software. 

Well, I'll take a look to see if there is something fishy going on.  What version are you using on your rig currently?  

> I'm just glad I found a workaround as my wife was starting to question if it would've been cheaper to by the wifi Traeger model. :-( Now I can tell her that definitely wouldn't have been cheaper and I have more functionality. :-)

Ha!  Well, this version is also way more fun to tinker with.  :)   But my wife would probably also be frustrated with me if the grill was out of commission for more than a weekend. :P I should get a test grill to play with so that my actual production grill doesn't have down time.

**Comment by plazarch** [comment]
PiFire Version: 1.3.4

**Comment by plazarch** [comment]
So, I thought I had it licked.  Installed everything today, after a few wiring setbacks, I thought I had everything working.  Bench tested the unit and everything was working.  Ignitor, fan, auger, so I put everything back in the grill.  Powered it up, set the grill probe to OEM and it's back to it's old tricks:
![PXL_20220827_231244082-1](https://user-images.githubusercontent.com/16919300/187051588-7c600b6b-d34a-49d4-b16b-1fb6d652efb1.jpg)
I know the probe is good so I'm hanging it up for today.  If there is any kind of debug script I can run on it, I would be happy to, just point me to it.

**Reply by nebhead** [reply]
Oh bummer!  Maybe try to go back to the original probe sources again?  

```json
    "probe_sources": [
      "ADC0",
      "ADC1",
      "ADC2"
    ],
```
Also, if you can show me what you have set for probe types:

```json
  "probe_types": {
    "grill0type": "PT1000-00",
    "grill1type": "PT1000-00",
    "grill2type": "TWPS00",
    "probe1type": "TWPS00",
    "probe2type": "TWPS00"
  },
```

If none of this helps, we can try to go and look at your probe tuning, using the tuning tool.  This may require you to fire up your grill, or get some cold water, warm water and hot water to do some tuning with the probe.  

![Screenshot from 2022-08-27 17-00-00](https://user-images.githubusercontent.com/19876886/187051958-d1432d14-9a0b-41b9-9ccd-a19606b6e5c6.png)

![Screenshot from 2022-08-27 17-01-10](https://user-images.githubusercontent.com/19876886/187051993-48d836f9-2fd3-4608-be5f-c1fb16aacd2c.png)

You can create a new probe profile with the coefficients that are generated when you do the tuning.  This might be needed if your probe is different than the PT1000 probe that Traeger uses.  It's also just nice to dial in your setup so that the probe is more accurate.   

If all of that doesn't seem to work, we might need to go back to see what version of the ADC python module you have, and start doing some more software debug. 

We will get to the bottom of this, I promise!

**Comment by plazarch** [comment]
So, none of that helped.  Here is the relevent info from settings.json around probes, the python module version, and system info:
```json  "grill_probe_settings": {
    "grill_probe": "grill_probe1",
    "grill_probe_enabled": [
      1,
      0,
      0
    ],
    "grill_probes": {
      "grill_probe1": {
        "name": "Grill Probe 1"
      },
      "grill_probe2": {
        "name": "Grill Probe 2"
      },
      "grill_probe3": {
        "name": "Avg Grill Probes"
      }
    }


  "modules": {
    "adc": "ads1115",
    "display": "ili9341b",
    "dist": "vl53l0x",
    "grillplat": "pifire"
  },

  "probe_settings": {
    "probe_profiles": {
      "ET73-HM": {
        "A": 0.00024723753,
        "B": 0.00023402251,
        "C": 1.3879768e-07,
        "Rd": 10000,
        "Vs": 3.28,
        "name": "ET-73-Heatermeter"
      },
      "ET73-SP": {
        "A": 0.00023067434,
        "B": 0.00023696596,
        "C": 1.2636414e-07,
        "Rd": 10000,
        "Vs": 3.28,
        "name": "ET-73-skyeperry1"
      },
      "My-OEM-Grill-Probe": {
        "A": 0.1204603049239879,
        "B": -0.023242792768715236,
        "C": 0.0001329908192874877,
        "Rd": 10000,
        "Vs": 3.28,
        "name": "My-OEM-Grill-Probe"
      },


    "probe_sources": [
      "ADC1",
      "ADC2",
      "ADC3"
    ],
    "probes_enabled": [
      1,
      1,
      1
    ]
  },
  "probe_types": {
    "grill1type": "My-OEM-Grill-Probe",
    "grill2type": "My-OEM-Grill-Probe",
    "probe1type": "TWPS00",
    "probe2type": "TWPS00"
  },
```
Python Module version:     ADS1115            0.2.1

System Info:
CPU Info
processor : 0
model name : ARMv6-compatible processor rev 7 (v6l)
BogoMIPS : 997.08
Features : half thumb fastmult vfp edsp java tls
CPU implementer : 0x41
CPU architecture: 7
CPU variant : 0x0
CPU part : 0xb76
CPU revision : 7

Hardware : BCM2835
Revision : 9000c1
Serial : 00000000e7d3c639
Model : Raspberry Pi Zero W Rev 1.1

PiFire Version: 1.3.4

**Thanks Again for all your help and your quick responses.**

**Reply by nebhead** [reply]
No problem!  Having a lazy Sunday over here, so this is good debug fun. 

OK, so when you have the following settings for probe sources:

```json
    "probe_sources": [
      "ADC1",
      "ADC2",
      "ADC3"
    ]
```

That means you'll be mapping the ADC outputs as follows: 

ADC0 -> Nothing
ADC1 -> Grill Probe
ADC2 -> Probe1
ADC3 -> Probe2

Where the PCB is routing the following:

ADC0 -> Grill Probe
ADC1 -> Probe1
ADC2 -> Probe2
ADC3 -> Nothing

So, I'm trying to figure out how you got it to work with the above settings previously.  

And just for sanity, can you check your resistors that you are using with a multimeter to make absolutely sure they are 10k?  The markings look suspiciously like they are 120ohm, but it just could be how the manufacturer marked them.  

![image](https://user-images.githubusercontent.com/19876886/187080162-493c48e3-8e7c-4c4a-b8be-58e0eda5d5b0.png)

Beyond that, all I can think of here is that your grill probe resistance is different than what we expect with your current probe profile, and is throwing things off with the calculations.  Maybe just measure the resistance at room temp with your multimeter across the grill probe to see what kind of resistance you are getting with that.  Also, what kind of grill probe are you using specifically?

Did you try the probe tuning tool with the older settings?  What kind of resistance values were you getting on the Grill probe at different temperatures?

**Comment by plazarch** [comment]
Ok, so when I set the probe sources with ADC0 first everything still shifts by 1.  In the displays both on the grill and in the web the grill temp will show as 0 degrees, probe 1 shows as 329 degrees (with nothing plugged in to probe 1) and probe 2 shows 0 degrees no matter if a probe is plugged in or not.  Removing ADC0 in the displays Grill shows 329 degrees and probe 1 and probe 2 work normally.

My first thought was that the resistors might be wrong, but they are reading 9.95k with the multimeter.  Yeah, the markings are confusing.

I'm using the original Traeger probe.

Here are the resistance settings:
 33 Deg F:  1.020k  ohms
 80 Deg F:  1.099k ohms.
212 Deg F: 1.362 k ohms

**Reply by nebhead** [reply]
OK, I think we are closing in on this.  Your resistances look good for that probe, so it should be working fine.  

Let's take a look at the ADS1115.  Can you take a picture of your module and post it here?  I feel like that might be part of the problem, since the mapping you have set indicates that something isn't right with how we are getting data from the ADC.  Since the channel 3 on the ADS1115 is a no-connect on the PCB you have, it should just be floating.  

Also, just to make sure we have the right modules installed, let's check your versions:

In the terminal do:
```bash
python3 -m pip list 
```
and send the output.

**Reply by plazarch** [reply]
Here is the pip list.  It's going to take me a little bit to get everyting unplugged and get pictures. 
Package            Version
------------------ ---------
ADS1115            0.2.1
async-timeout      4.0.2
bidict             0.22.0
cbor2              5.4.3
certifi            2020.6.20
chardet            4.0.0
ciso8601           2.2.0
click              8.1.3
colorzero          1.1
Deprecated         1.2.13
distro             1.5.0
dnspython          1.16.0
eventlet           0.30.2
Flask              2.2.2
Flask-Mobility     1.1.0
Flask-QRcode       3.1.0
Flask-SocketIO     5.3.0
gpiozero           1.6.2
greenlet           1.1.3
gunicorn           20.1.0
idna               2.10
importlib-metadata 4.12.0
influxdb-client    1.32.0
itsdangerous       2.1.2
Jinja2             3.1.2
luma.core          2.3.1
luma.lcd           2.9.0
luma.oled          3.8.1
MarkupSafe         2.1.1
olefile            0.46
packaging          21.3
Pillow             9.1.1
pip                20.3.4
pushbullet.py      0.12.0
pyftdi             0.54.0
pyparsing          3.0.9
pyserial           3.5
python-apt         2.2.1
python-dateutil    2.8.2
python-engineio    4.3.4
python-magic       0.4.27
python-socketio    5.7.1
pyusb              1.2.1
qrcode             7.3.1
reactivex          4.0.4
redis              4.3.4
requests           2.25.1
RPi.GPIO           0.7.0
setuptools         52.0.0
six                1.16.0
smbus2             0.4.2
spidev             3.5
ssh-import-id      5.10
supervisor         4.2.2
typing-extensions  4.3.0
urllib3            1.26.5
uuid               1.30
VL53L0X            1.0.4
websocket-client   1.4.0
Werkzeug           2.2.2
wheel              0.34.2
wrapt              1.14.1
zipp               3.8.1

**Comment by plazarch** [comment]
Just as an fyi, I temporari;y enable the 2 print lines in adc_ads1115.py:
                        # print('Probe: ' + probe_profile['name'])
                        # print('Probe Resistance: ' + str(Tr) + '\n')
and here are a couple of the lines from control.out.log with the probe at room temp ~ 80 deg F:
Probe: My-OEM-Grill-Probe
Probe Resistance: 5295.697198362367

Probe: Thermoworks-Pro-Series-HeaterMeter
Probe Resistance: 1116.5243915355124

Probe: My-OEM-Grill-Probe
Probe Resistance: 5346.197092603004

Probe: Thermoworks-Pro-Series-HeaterMeter
Probe Resistance: 1112.9933931899036

Probe: My-OEM-Grill-Probe
Probe Resistance: 5238.94293451008

**Comment by plazarch** [comment]
I'm beginning to wonder if the ADS1115 is either defective or just a complete piece of Chinese crap. :-(
Although I was getting the same type of readings with a completely different ADS.  It was a full hat type ADS.

**Reply by nebhead** [reply]
Well, I have a hunch that it's routed differently and that is why things aren't adding up.  It does look like it's working where it's connected to the right channels.

**Comment by plazarch** [comment]
Here's front and back images of the ADS1115 Module:
![PXL_20220828_161906270](https://user-images.githubusercontent.com/16919300/187084329-92d410bb-d452-4d76-832a-c6763dd330a4.jpg)
![PXL_20220828_161915257](https://user-images.githubusercontent.com/16919300/187084332-6e5f5aea-da81-4be3-818b-47272574ae75.jpg)

**Reply by nebhead** [reply]
Hmm... that looks to be a bog standard ADS1115 with the right pin-out.  Maybe I'm barking up the wrong tree there.  Then it's got to be the mapping in PiFire.  I think I need to go back and look at the logic again.  

Maybe if you can send your entire settings.json file, but strip out any personal information (i.e. API keys) if you have any.  I'll start there.

**Reply by plazarch** [reply]
Here you go:
```json
{
  "cycle_data": {
    "HoldCycleTime": 20,
    "PB": 60.0,
    "PMode": 2,
    "SmokeCycleTime": 15,
    "Td": 45.0,
    "Ti": 180.0,
    "center": 0.5,
    "u_max": 1.0,
    "u_min": 0.15
  },
  "globals": {
    "augerrate": 0.3,
    "auto_power_off": false,
    "buttonslevel": "HIGH",
    "debug_mode": false,
    "disp_rotation": 0,
    "first_time_setup": false,
    "four_probes": false,
    "grill_name": "Traeger Pro 22",
    "page_theme": "dark",
    "shutdown_timer": 60,
    "startup_timer": 240,
    "triggerlevel": "HIGH",
    "units": "F"
  },
  "grill_probe_settings": {
    "grill_probe": "grill_probe1",
    "grill_probe_enabled": [
      1,
      0,
      0
    ],
    "grill_probes": {
      "grill_probe1": {
        "name": "Grill Probe 1"
      },
      "grill_probe2": {
        "name": "Grill Probe 2"
      },
      "grill_probe3": {
        "name": "Avg Grill Probes"
      }
    }
  },
  "history_page": {
    "autorefresh": "on",
    "clearhistoryonstart": true,
    "datapoints": 60,
    "minutes": 15
  },
  "ifttt": {
    "APIKey": "",
    "enabled": false
  },
  "influxdb": {
    "bucket": "",
    "enabled": false,
    "org": "",
    "token": "",
    "url": ""
  },
  "inpins": {
    "selector": 17
  },
  "keep_warm": {
    "s_plus": false,
    "temp": 165
  },
  "lastupdated": {
    "time": 1661700781
  },
  "modules": {
    "adc": "ads1115",
    "display": "ili9341b",
    "dist": "vl53l0x",
    "grillplat": "pifire"
  },
  "onesignal": {
    "app_id": "99999999",
    "devices": {
      "99999999": {
        "app_version": "1.4.3",
        "device_name": "xxxxxx",
        "friendly_name": ""
      }
    },
    "enabled": false,
    "uuid": "99999999"
  },
  "outpins": {
    "auger": 14,
    "fan": 15,
    "igniter": 18,
    "power": 4
  },
  "pelletlevel": {
    "empty": 22,
    "full": 4,
    "warning_enabled": true,
    "warning_level": 25
  },
  "probe_settings": {
    "probe_profiles": {
      "ET73-HM": {
        "A": 0.00024723753,
        "B": 0.00023402251,
        "C": 1.3879768e-07,
        "Rd": 10000,
        "Vs": 3.28,
        "name": "ET-73-Heatermeter"
      },
      "ET73-SP": {
        "A": 0.00023067434,
        "B": 0.00023696596,
        "C": 1.2636414e-07,
        "Rd": 10000,
        "Vs": 3.28,
        "name": "ET-73-skyeperry1"
      },
      "My-OEM-Grill-Probe": {
        "A": 0.1204603049239879,
        "B": -0.023242792768715236,
        "C": 0.0001329908192874877,
        "Rd": 99500,
        "Vs": 3.28,
        "name": "My-OEM-Grill-Probe"
      },
      "PT-1000-OEM": {
        "A": 0.04136906456,
        "B": -0.00677987613,
        "C": 2.760294589e-05,
        "Rd": 10000,
        "Vs": 3.28,
        "name": "PT-1000-Grill-Probe-OEM"
      },
      "PT-1000-PiFire": {
        "A": 0.05469905897345206,
        "B": -0.009473055040089443,
        "C": 4.3768560703857386e-05,
        "Rd": 10000,
        "Vs": 3.28,
        "name": "PT-1000-Grill-Probe-PiFire"
      },
      "TWPS00": {
        "A": 0.00073431401,
        "B": 0.0002157437,
        "C": 9.515686e-08,
        "Rd": 10000,
        "Vs": 3.28,
        "name": "Thermoworks-Pro-Series-HeaterMeter"
      },
      "iGrill-HM": {
        "A": 0.0007739251279,
        "B": 0.0002088025997,
        "C": 1.154400438e-07,
        "Rd": 10000,
        "Vs": 3.28,
        "name": "iGrill-Heatermeter"
      }
    },
    "probe_sources": [
      "ADC1",
      "ADC2",
      "ADC3"
    ],
    "probes_enabled": [
      1,
      1,
      1
    ]
  },
  "probe_types": {
    "grill1type": "My-OEM-Grill-Probe",
    "grill2type": "My-OEM-Grill-Probe",
    "probe1type": "TWPS00",
    "probe2type": "TWPS00"
  },
  "pushbullet": {
    "APIKey": "",
    "PublicURL": "",
    "enabled": false
  },
  "pushover": {
    "APIKey": "",
    "PublicURL": "",
    "UserKeys": "",
    "enabled": false
  },
  "safety": {
    "maxstartuptemp": 100,
    "maxtemp": 550,
    "minstartuptemp": 75,
    "reigniteretries": 1
  },
  "smartstart": {
    "enabled": true,
    "profiles": [
      {
        "augerontime": 15,
        "p_mode": 0,
        "startuptime": 360
      },
      {
        "augerontime": 15,
        "p_mode": 1,
        "startuptime": 360
      },
      {
        "augerontime": 15,
        "p_mode": 3,
        "startuptime": 240
      },
      {
        "augerontime": 15,
        "p_mode": 5,
        "startuptime": 240
      }
    ],
    "temp_range_list": [
      60,
      80,
      90
    ]
  },
  "smoke_plus": {
    "cycle": 10,
    "duty_cycle": 50,
    "enabled": false,
    "frequency": 1,
    "max_temp": 220,
    "min_temp": 160
  },
  "versions": {
    "server": "1.3.4"
  }
}

**Comment by nebhead** [comment]
I can't see anything wrong with your settings, and I'm trying to see where the logic would have gone awry.  And now I'm back to wondering if the ADC you have is wonky.  It's like it's returning channel values in the wrong order. 

Let's try to swizzle the ADC probe settings as follows:
```json
    "probe_sources": [
      "ADC3",
      "ADC2",
      "ADC1"
    ],
```

**Comment by plazarch** [comment]
That changes the Grill Probe to Display as Probe 2 and does something really wonky to the display:
![Wonky_Screenshot](https://user-images.githubusercontent.com/16919300/187086959-0aed19cb-0caf-47cb-8667-28f7f1f063c9.png)

**Reply by nebhead** [reply]
Oh, yes, that's kind of expected for the display as it's doing the calculations with a maximum temperature value of 300F.  I should probably bounds check that before displaying.  

While we have it in this config, try plugging in Probe 1, then Probe 2, then both and observe what happens. 

The ADC is really behaving strangely.  I am wondering if it's defective or if the manufacturer put a different ADC component on the board to save cost.  [One of the comments on Amazon indicated as much.](https://www.amazon.com/product-reviews/B082QX1RDV/ref=acr_dp_hist_1?ie=UTF8&filterByStar=one_star&reviewerType=all_reviews#reviews-filter-bar) I hate to say it, but maybe it's time to order a new one?

**Comment by plazarch** [comment]
I just ordered a replacement set of ADS1115's from Amazon.  They should be delivered Wednesday.

**Reply by nebhead** [reply]
Yep, probably a good call.  I didn't want you to have to spend any more money on this, but we might need to in this case.

**Comment by plazarch** [comment]
With that config, plugging in probe 1 gives me a normal reading on Probe 1 on the dashboard, nothing on Grill temp, and 328 on Probe 2.  Plugging in to Probe 2 Gives 0 on Grill, 0 on Probe 1 and 328 on Probe 2.  I can't plug both probes in until tomorrow as the second jack hasn't been delivered yet.  I think it's time to step back and wait for the new ADS1115's.

**Reply by nebhead** [reply]
Yep, lets wait for that new one to come in.  Then, if needed, I think I should write a quick test script to just directly read the channels of the ADC and see what is going on.

**Comment by plazarch** [comment]
Well. after a long wait for Amazon to get there **** together I finally received the new ADS's today.  Assembled one and ended up with the exact same issue.  I wrote a couple quick scripts to gather voltage readings from the ADS, one using the ADC1115 module and one using the adafruit_ads1x15 module.  While readings are fairly consistent from reading to reading in each module there is a huge difference between the two modules:
![ADS_Measurments](https://user-images.githubusercontent.com/16919300/188236584-9e5d0e7e-3dfa-4b74-8a7c-ceda415b48ec.jpg)

**Reply by nebhead** [reply]
Hmm... At least we are getting some results that look like they are consistent.  Let me draft up a script you can use and see if we can figure out what else might be going on here.

**Reply by nebhead** [reply]
OK, just ran a the super simple script on my instance of pifire and here is what I'm getting:

```python3
import ADS1115
import time

ads = ADS1115.ADS1115()

adc_value = [0,0,0,0]

for index in range(4):
    time.sleep(0.05)
    adc_value[index] = ads.readADCSingleEnded(index)
    print(f'adc[{index}] = {adc_value[index]}')
```

And the results are below, reporting in mV using the ADS1115 module of course.

Run 1:
```
adc[0] = 341.25  <= Traeger OEM Grill Probe PT-1000
adc[1] = 2771.25 <= Thermoworks Pro Series
adc[2] = 2770.5 <= Thermoworks Pro Series
adc[3] = 554.625 <= Unconnected
```

Run 2:
```
adc[0] = 341.0625
adc[1] = 2757.0
adc[2] = 2772.5625
adc[3] = 552.5625
```
Run 3:
```
adc[0] = 341.0625
adc[1] = 2757.0
adc[2] = 2772.5625
adc[3] = 552.5625
```

Maybe you can try the above code on yours with the probes connected and see how it goes.  Right now, it looks like there is an issue with the ADS1115 python module that you have installed.  I think that is going to be the culprit.  Though I cannot quite understand how that could be the case.

**Comment by plazarch** [comment]
adc[0] = 4415.25        <= Traeger OEM Grill Probe PT-1000
adc[1] = 4410.9375      <= Thermoworks Pro Series
adc[2] = 2942.8125      <= Thermoworks Pro Series
adc[3] = 2940.1875      <= Nothing

adc[0] = 196.5
adc[1] = 3297.9375
adc[2] = 4950.75
adc[3] = 2940.5625

adc[0] = 195.5625
adc[1] = 3298.5
adc[2] = 4415.0625
adc[3] = 2940.5625

**Reply by nebhead** [reply]
Whoa!  We've got values above 3.3V's there, which would be impossible (unless your input voltage was 5V, which it shouldn't be).  And the values are all over the map.  Can you send me the code you wrote for the Adafruit test?  I still keep thinking there must be some hardware change that the ADS1115 python module isn't interpreting right.

**Comment by plazarch** [comment]
I've been playing with it a bit, here is the output of the latest run:
Port	 Volts
 0 	0.1314
 1 	2.9443
 2 	2.9417
 3 	2.9440
Port	 Volts
 0 	2.9416
 1 	0.1328
 2 	2.9417
 3 	0.1320
Port	 Volts
 0 	2.9448
 1 	2.9413
 2 	0.1315
 3 	0.1316
Port	 Volts
 0 	3.2982
 1 	2.9443
 2 	2.9418
 3 	0.1290
Port	 Volts
 0 	3.2997
 1 	2.9446
 2 	2.9448
 3 	2.9417
and here is the script:
```
import board
import busio
import time
import datetime
from Color import *

i2c = busio.I2C(board.SCL, board.SDA)

import adafruit_ads1x15.ads1115 as ADS

from adafruit_ads1x15.analog_in import AnalogIn

ads = ADS.ADS1115(i2c)
pins = [ADS.P0, ADS.P1, ADS.P2, ADS.P3]

adc_value = [[],[],[],[]]
colors = [Color.F_Red, Color.F_Green, Color.F_Blue, Color.F_Yellow]

i = 0

while i < 5:
    try:
        for index in range(4):
            time.sleep(0.05)
            adc_value[index] = AnalogIn(ads, pins[index])
            if index == 0:
                print("Port\t", "Volts")
            print(colors[index], str(index), "\t{:.4f}".format(adc_value[index].voltage), Color.F_Default)
    except Exception as inst:
        now = str(datetime.datetime.now())
        now = now[0:19] # Truncate the microseconds
        print(str(now) + ' Error Reading Temperature.: ' + inst)

    i += 1
```
I got the modules and example from here: https://learn.adafruit.com/adafruit-4-channel-adc-breakouts/python-circuitpython

**Reply by nebhead** [reply]
I built a super simple version of the script here:
```python3
import time
import board
import busio
import adafruit_ads1x15.ads1015 as ADS
from adafruit_ads1x15.analog_in import AnalogIn

# Create the I2C bus
i2c = busio.I2C(board.SCL, board.SDA)

# Create the ADC object using the I2C bus
ads = ADS.ADS1015(i2c)

channel = [0,0,0,0]
# Create single-ended input on channel 0
channel[0] = AnalogIn(ads, ADS.P0)
channel[1] = AnalogIn(ads, ADS.P1)
channel[2] = AnalogIn(ads, ADS.P2)
channel[3] = AnalogIn(ads, ADS.P3)

# Create differential input between channel 0 and 1
#chan = AnalogIn(ads, ADS.P0, ADS.P1)

for index in range(0, len(channel)):
	print(f'ADC[{index}] = {channel[index].voltage * 1000}')
	time.sleep(0.5)
```
Then I ran it few times with the probes plugged in:

```
ADC[0] = 334.01019318216504
ADC[1] = 2862.0873439741204
ADC[2] = 2876.087771233253
ADC[3] = 562.0171514023256
pi@pifire:~/pifire $ python3 2test.py 
ADC[0] = 334.01019318216504
ADC[1] = 2862.0873439741204
ADC[2] = 2876.087771233253
ADC[3] = 562.0171514023256
pi@pifire:~/pifire $ python3 2test.py 
ADC[0] = 334.01019318216504
ADC[1] = 2862.0873439741204
ADC[2] = 2876.087771233253
ADC[3] = 562.0171514023256
```

So, I'm getting ~562mv on the unconnected input, so that lines up with your HAT testing.  Though, since there is a 10k Ohm resistor on the PiFire board, it should be reading closer to 3300mV (or 3.3V) on ADC0, 1, 2 if they aren't connected to a probe.  

Also, on your runs above this comment, I'm assuming with the Adafruit module, you are getting lots of varying input on different ports.  So yeah, there looks to be some kind of hardware issue going on.

**Comment by plazarch** [comment]
Ok, I decided to step back and do some basic ground up testing.  I took everything off the board and brought it back inside.  I plugged in my hat type ads1115 and ran my adafruit type script and got roughly 0.57 volts on each port with no probes plugged in.  I then shut down, removed the hat type ads1115 and plugged in the small ads1115 into the main board and ran the adafruit script again and it gives me roughly 3.2 volts on ports 0 - 3 and 0.12 volts on port 3 so there is definitely something wonky with the hardware.

**Reply by nebhead** [reply]
> Ok, I decided to step back and do some basic ground up testing. I took everything off the board and brought it back inside. I plugged in my hat type ads1115 and ran my adafruit type script and got roughly 0.57 volts on each port with no probes plugged in. 

It sounds like these connections are not using the PiFire board, right?  These are unconnected to anything, if I'm not mistaken. 

>I then shut down, removed the hat type ads1115 and plugged in the small ads1115 into the main board and ran the adafruit script again and it gives me roughly 3.2 volts on ports 0 - 3 and 0.12 volts on port 3 so there is definitely something wonky with the hardware.

This actually sounds closer to right.  Without any probes connected, it would be 3.2-3.3V for each of the connected ports.  Port3 would be something like 0.5V, but 0.12V might not be too far off.

**Comment by plazarch** [comment]
I unplugged the pi from the mainboard and used jumper wires to connect the ADS directly to the pi and re-ran the script again with noting plugged in:
Port	 Volts
 0 	0.1181
 1 	0.1223
 2 	0.1216
 3 	0.1184
Time to start probing the main board and see where the crossed wires are.

**Reply by nebhead** [reply]
Yeah, there is something weird here.  I noticed you had some dupont connectors on your probe wires too.  That might be a source of difficulty, because those connections could be unstable.  I'd say remove those connectors in the middle and see if that helps.  Inputs to the ADC can be very sensitive.

**Comment by plazarch** [comment]
Yesterday I came to the conclusion that there may have been a software issue, so I wiped the sd card, reimaged it, ran the os configs and applied all updates.  Then I installed just the ADS1115 python module and ran the test script several times.  I couldn't get stable correct readings.  So then I installed the Adafruit ADS1x15 module and ran the Adafruit test script many times and received very stable readings.  I then installed PiFire and configured everything and in the stock form I couldn't get stable readings.  In stock form the ADS1115 script was putting the reading from ADC0 into the result for ADC1 and putting nonsense values into the result for ADC0, and the same thing was happing inside PiFire.  I modified adc_ads1115.py to use the Adafruit module and I get stable results.  The grill probe is showing roughly the current room temp but the Thermoworks probe is way off.  I am waiting for the wife to finish with the oven so I can tune the probes and see what happens.  My fingers, toes and eyes are crossed that everything will work. :-)  I wish I knew exactly what was wrong with the ADS1115 module, but if this works I'll be happy with it.  Thank you for all your time and effort in resolving this and I'll keep you updated with the final results.

**Reply by nebhead** [reply]
Well, best of luck on this part.  I hope this works out.  If it does, perhaps it would be a good idea to add a new pifire ADC module using the Adafruit module as a variant in case others run into a similar issue.  I just keep thinking there must have been some change in hardware that the old ADS1115 module doesn't seem to recognize.  So maybe it's just a good idea to move to the adafruit version, which is maintained.  The ADS1115 module that we are currently using was last updated in January of 2019.

**Comment by plazarch** [comment]
Round one with the Probe Tuning and the Grill probe appears to be working well.  The Thermopro is closer but still a ways off and I had to give the oven back.  :-)  I'm making baby steps so I hope to be smoking again soon!

**Comment by nebhead** [comment]
Nice!  The most important one is the grill probe, so you are mostly there!

**Reply by nebhead** [reply]
Thanks for the donation!  I'm hoping for the best when you are able to pick it up again.  I may switch up the ADS1115 module for the Adafruit in the official release to ensure it's robust.  Or perhaps just add a new module. Anyway, best of luck!

**Comment by plazarch** [comment]
I guess I haven't walked far enough away from this yet. :-)  I did get the original controller back in so I can cook for now.  I thought you might like to look over what I came up with for the Adafruit module.  Maybe you can spot anything I did wrong or maybe it will give you a head start.  You can download it here.  In my setup I just called it adc_ads1115.py so I didn't have to make any other changes to the install or update modules for now.  https://www.vanzarsystems.com/index.php/downloads/3-pifire?download=3:adc-ads1x115

**Reply by nebhead** [reply]
Took a quick peek and it looks OK from what I see.  Definitely could give me a heads start, or be the solution.  If you wanted to, you could put up a pull request against the development branch.  Otherwise, I'm happy to push it up myself.

**Comment by plazarch** [comment]
I've never done a pull request, but I would love to learn.  I started and it looks pretty much like an issue, but it won't let me upload a .py file.  Should I just paste the contents of the .py file into the body of the pull request?

**Reply by nebhead** [reply]
Sure!  What you'll want to do first, is fork the project on GitHub.  When you've done that, you can clone your new forked repository.  Make sure that on your cloned local copy, you checkout the development branch, then go ahead and add your ads1x15.py file to the repository.  Make a git commit to your fork, adding this new file.  Then git push this change back to GitHub.  From GitHub, you can raise a pull-request against the original project (there should be buttons on your forked project).  From there, I will review and accept the changes.  

But... before you do that I think I have some feedback on changes since I see at least one possible bug.

**Reply by nebhead** [reply]
OK, so I'm not sure if this is a bug or not, since I haven't tried this specifically, but on this line:

```python3
adc_value[index] = math.floor(readData.voltage * 100)
```
To ensure you are getting the right values should be multiplied by 1000 to get the proper mV value from the Adafruit module readings.  

```python3
adc_value[index] = math.floor(readData.voltage * 1000)
```

And likewise with the conversion, make sure you change:

```python3
if(adc_value > 0) and (adc_value < (probe_profile['Vs'] * 100) * 0.99):
```

to

```python3
if(adc_value > 0) and (adc_value < (probe_profile['Vs'] * 1000) * 0.99):
```

**Reply by plazarch** [reply]
I actually had to set it to 100 instead of 1000 to get usable readings.  I had the same issue with the test scripts that the original ADS module needed * 1000 and the Adafruit module needed * 100 to produce the same values.

**Reply by plazarch** [reply]
I just checked and as long as it is the same in both places you end up with the same readings.  I am updating to use the 1000 numbers just to keep things consistent.

**Reply by nebhead** [reply]
Great, thanks.  The Adafruit values should be in Volts where the conversion function is expecting millivolts.  Just want to make sure we remove the possibility for other potential bugs / differences from the original code.

**Reply by nebhead** [reply]
Congrats on your first successful pull request!  You are now immortalized as a contributor to this project.  Once we have a decent amount of testing with the development branch, I'll push the changes to the main with the next update.

**Comment by plazarch** [comment]
I missed you last line about the possible bug. :-(  The pull request is done.  I assume you can reject it and I can re-do one later.
